examples/pytorch/rgcn/entity_classify_mp.py

```diff
@@ -269,54 +308,107 @@ def run(proc_id, n_gpus, args, devices, dataset):
 
         # only process 0 will do the evaluation
-        if proc_id == 0:
+        if (queue is not None) or (proc_id == 0):
             model.eval()
-            eval_logtis = []
+            eval_logits = []
             eval_seeds = []
-            for i, sample_data in enumerate(val_loader):
-                seeds, blocks = sample_data
-                feats = embed_layer(blocks[0].srcdata[dgl.NID].to(dev_id),
-                                    blocks[0].srcdata[dgl.NTYPE].to(dev_id),
-                                    node_feats)
-                logits = model(blocks, feats)
-                eval_logtis.append(logits)
-                eval_seeds.append(seeds)
-            eval_logtis = th.cat(eval_logtis)
-            eval_seeds = th.cat(eval_seeds)
-            val_loss = F.cross_entropy(eval_logtis, labels[eval_seeds])
-            val_acc = th.sum(eval_logtis.argmax(dim=1) == labels[eval_seeds]).item() / len(eval_seeds)
-            print("Validation Accuracy: {:.4f} | Validation loss: {:.4f}".
-                    format(val_acc, val_loss.item()))
+            with th.no_grad():
+                for sample_data in tqdm.tqdm(val_loader):
+                    th.cuda.empty_cache()
+                    seeds, blocks = sample_data
+                    if args.mix_cpu_gpu is False:
+                        feats = embed_layer(blocks[0].srcdata[dgl.NID],
+                                            blocks[0].srcdata[dgl.NTYPE],
+                                            blocks[0].srcdata['type_id'],
+                                            node_feats)
+                    else:
+                        feats = embed_layer(blocks[0].srcdata[dgl.NID],
+                                            blocks[0].srcdata[dgl.NTYPE],
+                                            blocks[0].srcdata['type_id'],
+                                            node_feats)
+                    logits = model(blocks, feats)
+                    eval_logits.append(logits.cpu().detach())
+                    eval_seeds.append(seeds.cpu().detach())
+                eval_logits = th.cat(eval_logits)
+                eval_seeds = th.cat(eval_seeds)
+                if queue is not None:
+                    queue.put((eval_logits, eval_seeds))
+
+            if proc_id == 0:
+                if queue is not None:
+                    eval_logits = []
+                    eval_seeds = []
+                    for i in range(n_gpus):
+                        log = queue.get()
+                        val_l, val_s = log
+                        eval_logits.append(val_l)
+                        eval_seeds.append(val_s)
+                    eval_logits = th.cat(eval_logits)
+                    eval_seeds = th.cat(eval_seeds)
+                val_loss = F.cross_entropy(eval_logits, labels[eval_seeds].cpu()).item()
+                val_acc = th.sum(eval_logits.argmax(dim=1) == labels[eval_seeds].cpu()).item() / len(eval_seeds)
```

Comment:
it's better to put the evaluation code in a function, so it can be reused. also make the code more readable.

Reply:
Yes, validation and test can use same pieces of code.